cobra/model_building/models.py

```diff
@@ -163,39 +165,97 @@ def evaluate(self, X: pd.DataFrame, y: pd.Series,
+                                  average_return_for_successful_letter=75.0)
+            Any metric function you provide here should be a function taking
+            y_true, y_pred and/or y_score arguments, of numpy array type,
+            and optionally also additional arguments, which you can pass
+            through the metric_args and metric_kwargs parameters.
+            If you are unsure which arguments of your metric function are
+            args/kwargs, then run inspect.getfullargspec(your_metric_function).
+        metric_args : dict, optional
+            Arguments (for example: lift_at=0.05) to be passed to the metric
+            function when evaluating the model's performance.
+            Example metric function in which this is required:
+            ClassificationEvaluator._compute_lift(y_true=y_true,
+                                                  y_score=y_pred,
+                                                  lift_at=0.05)
+        metric_kwargs : dict, optional
+            Keyword arguments (for example: normalize=True) to be passed to the
+            metric function when evaluating the model's performance.
+            Example metric function in which this is required (from
+            scikit-learn):
+            def accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None)
 
         Returns
         -------
         float
             The performance score of the model (AUC by default).
         """
-        if metric is not None:  # decouple from _eval_metrics_by_split attribute
-            y_pred = self.score_model(X)
+        y_score = self.score_model(X)
 
-            fpr, tpr, thresholds = roc_curve(y_true=y, y_score=y_pred)
-            cutoff = (ClassificationEvaluator._compute_optimal_cutoff(fpr, tpr, thresholds))
-            y_pred_b = np.array([0 if pred <= cutoff else 1 for pred in y_pred])
-
-            performance = metric(y_true=y, y_pred=y_pred_b)
+        if metric is None:
+            # No custom evaluation metric was chosen. We use AUC as default
+            # evaluation metric:
+            performance = roc_auc_score(y_true=y, y_score=y_score)
 
+        else:
+            # A custom evaluation metric was chosen. With the default
+            # metric AUC, the performance could be scored over all possible
+            # thresholds and based on y_score;
+            # now, with any evaluation metric possibly being used, y_pred may
+            # be required instead of y_score, which requires determining the
+            # optimal threshold first and then calculating y_pred.
+            fpr, tpr, thresholds = roc_curve(y_true=y, y_score=y_score)
+            cutoff = ClassificationEvaluator._compute_optimal_cutoff(fpr, tpr,
+                                                                     thresholds)
+            y_pred = np.array([0 if score <= cutoff
+                               else 1
+                               for score in y_score])
+
```

Comment:
> If I understand well, if the metric function doesn't need y_pred, these calculations are redundant? Maybe try to catch the metric functions with an y_pred argument before running this.

Good thinking! I've added that.

> I also don't see y_proba documented, is this needed?

I noticed two of [sklearn's metric functions](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) use that argument name (actually y_prob instead of y_proba, modified that) rather than y_score and added it for the sake of completeness. I've added this explanation in the docstring: "The variables that we provide and your function can possibly take in are y_true, y_pred, y_score and y_prob."